### ISSUE TITLE
Pyright allows `|=` mutation of `collections.abc.Set`

### ISSUE THREAD

**Issue by ottumm** [microsoft/pyright#9737]
I would expect the following ([pyright playground](https://pyright-play.net/?code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSgGUlUAoFgEyWCmDDAAocAXPUYBtHKhABdAJRQAtAD4oAOWxIBTKFqg4oAHwC8UAN4AibmFMBfJkA)) to report an error because `Set` is immutable, but Pyright 1.1.392 reports it as clean:

```python
from collections.abc import Set


def foo(s: Set[str]) -> None:
    s |= {"foo"}
```

**Comment by erictraut** [comment]
Pyright has no specialized knowledge of `Set`. It bases is analysis on the type information provided in the bundled typeshed stubs. `Set` is an alias for the `typing.AbstractSet` class. It provides an `__or__` method, which means that the `|=` operator is supported. The runtime effectively changes the `s |= {"foo"}` statement into the operation `s = s | {"foo"}` which constructs a new object.

Not surprisingly, mypy likewise emits no errors for the above code.

In any case, pyright is working as intended here, so this isn't a pyright bug. If you think this should produce an error, please report the issue to the maintainers of typeshed.

**Comment by ottumm** [comment]
I see what you're saying.

In the codebase I work in, we use `collections.abc.Set` in function signatures to indicate that the function cannot mutate the input set (regardless of whether the concrete type is `set` or `frozenset`). I was surprised to find that this didn't hold for `|=`.

I'll take this up with typeshed, but to me it would make sense to remove `__ior__` from `collections.abc.Set`. It seems like anything that has the potential to mutate the object should only be available in `MutableSet`.

**Comment by erictraut** [comment]
The `collections.abc.Set` protocol doesn't support `__ior__`, so there's nothing to remove here. The `__ior__` magic method isn't required to support the `|=` operator. If `|=` is applied to an object that doesn't support `__ior__` but does support `__or__`, then the runtime calls the latter and assigns the resulting value (which is a new object) to the LHS expression. So `|=` can be used with classes that are immutable such as `frozenset`.
